Fix #3605: Introduce Platform Parameter WorkManager classes

Arjupta opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Fixes #3605 
Introduced a Lightweight Syncing mechanism for PlatformParameter with the help of Work Manager. This PR was stacked over https://github.com/oppia/oppia-android/pull/3489

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

Arjupta committed:
Added GaeModels + PlatformParameterService and its Mock

Arjupta committed:
Trying to fix bazel check by adding dependency over data module

Arjupta committed:
Changed the location of MockService to testing/network and of ServiceTest to data

Arjupta committed:
Bazel changes for TestConstants alomng with moving these constants into new file

Arjupta committed:
Nit change in the bazel package name

Arjupta committed:
Adding platform parameter BUILD.bazel as a dependency in testing

Arjupta committed:
Adding test_constants in testing/network BUILD.bazel and utility in testing/platformparameter BUILD.bazel

Arjupta committed:
Suggested changes related to exemption file, renaming test constants and Gae models also adding test cases in Service test and documentation over Gae models

Arjupta committed:
Nit change related to TODO issue number

Arjupta committed:
Reverting to the original response structure from backend. Removed GaeModels and used a Map<String,Any> instead

Arjupta committed:
Updating with develop

Arjupta committed:
Removed GaeModels from exemptions file & platform parameter from TEST_DEPS of testing module

Arjupta committed:
Nit change for renaming test function

Arjupta committed:
Nit change for adding named arguments

Arjupta committed:
Suggested change for testing request query, new files for test constants
and KDoc changes

Arjupta committed:
Static checks and updating exemption file

Arjupta committed:
Updating with develop

Arjupta committed:
Basic commit for introducing a Worker, WorkerFactory and WorkerInitializer

Arjupta committed:
Including all the files related to Platform Parameter Worker.

Arjupta committed:
SyncUp worker tests are running fine. Commented out Prod Platform Parameter Service Impl in Network Module

Arjupta committed:
Added the tests for SyncUpWorkManagerInitializer

Arjupta committed:
Introduced TestNetworkModule in the tests to remove the dependency from NetworkModule

Arjupta committed:
Modified exemption file to exclude Worker Factory and Module

Arjupta commented:
Just for clarification that as this PR is stacked one so it contains some commits which doesn't belong to this PR. Here are the commits which does not belong to this PR originally.

- [2ed701c](https://github.com/oppia/oppia-android/pull/3606/commits/2ed701ccce378bf43f0add0c734013b1aa812c0f)
- [6082592](https://github.com/oppia/oppia-android/pull/3606/commits/608259290d20d13007316757f04f175385c17003)
- [d354c05](https://github.com/oppia/oppia-android/pull/3606/commits/d354c0567fa8679ef5161bf4f8927161304058a5)

These are the first three commits as we can see right now. So any changes after these three, are the ones that originally belong to this PR.

Arjupta committed:
Added bazel dependency over PlatformParameterService and ContextExtensions

jcqli commented on domain/src/main/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorkManagerInitializer.kt:
```diff
@@ -0,0 +1,63 @@
+ */
+@Singleton
+class PlatformParameterSyncUpWorkManagerInitializer @Inject constructor(
+  private val context: Context,
```
nit: remove the comma at the end.

Arjupta replied:
Done

jcqli replied:
Not a relevant comment anymore since there's a second parameter.

jcqli commented on domain/src/main/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorkManagerInitializer.kt:
```diff
@@ -0,0 +1,63 @@
+    )
+    .build()
+
+  private val workRequestForSyncingUpPlatformParameters = PeriodicWorkRequest
```
`workRequestForSyncingPlatformParameters` might be easier to read.

Arjupta replied:
Done

jcqli commented on domain/src/main/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorkManagerInitializer.kt:
```diff
@@ -0,0 +1,63 @@
+  private val context: Context,
+) : ApplicationStartupListener {
+
+  private val OPPIA_PLATFORM_PARAMETER_WORK = "OPPIA_PLATFORM_PARAMETER_WORK_REQUEST"
```
Could probably name the val "OPPIA_PLATFORM_PARAMETER_WORK_REQUEST_NAME" since this is a String to identify / name the request, not the work.

Arjupta replied:
Done

jcqli commented on domain/src/main/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorker.kt:
```diff
@@ -0,0 +1,89 @@
+import org.oppia.android.util.threading.BackgroundDispatcher
+import javax.inject.Inject
+
+/** Worker class that fetches anc cache the latest platform parameters from the remote service. */
```
nit: `and caches` not `anc cache`

Arjupta replied:
Done

jcqli commented on domain/src/main/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorker.kt:
```diff
@@ -0,0 +1,89 @@
+        is String -> platformParameter.string = value
+        is Int -> platformParameter.integer = value
+        is Boolean -> platformParameter.boolean = value
+        else -> continue
```
Should an error be thrown here? Or, if it's supposed to be null so that we can set the parameter to default, add a comment explaining this.

Arjupta replied:
Done

jcqli commented on scripts/assets/test_file_exemptions.textproto:
```diff
@@ -555,6 +556,7 @@ exempted_file_path: "domain/src/main/java/org/oppia/android/domain/topic/Revisio
 exempted_file_path: "domain/src/main/java/org/oppia/android/domain/util/FloatExtensions.kt"
 exempted_file_path: "domain/src/main/java/org/oppia/android/domain/util/FractionExtensions.kt"
 exempted_file_path: "domain/src/main/java/org/oppia/android/domain/util/JsonAssetRetriever.kt"
+exempted_file_path: "domain/src/main/java/org/oppia/android/domain/workmanager/WorkManagerConfigurationModule.kt"
```
This is probably in alphabetical order so move this below `InteractionObjectTestBuilder`

Arjupta replied:
InteractionObjectTestBuilder has this import statement
**"domain/src/test/java/org/oppia/android/domain/classify/InteractionObjectTestBuilder.kt"**

whereas WorkManagerConfigurationModule has this 
**"domain/src/main/java/org/oppia/android/domain/workmanager/WorkManagerConfigurationModule.kt"**

jcqli replied:
Oh ah I see, thanks for clarifying!

jcqli reviewed (changes requested):
A couple comments but nothing major -- want to take a second pass on it later as well.

Arjupta committed:
Added NetworkModule in all the tests under where WorkManagerConfigurationModule is used

Arjupta committed:
Suggested nit changes

Arjupta committed:
Added tests for MockPlatformParameterService

Arjupta committed:
Updating with develop

Arjupta committed:
Updating with Base PR

vinitamurthi commented on domain/src/main/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorkManagerInitializer.kt:
```diff
@@ -0,0 +1,63 @@
+    .build()
+
+  private val workRequestForSyncingPlatformParameters = PeriodicWorkRequest
+    .Builder(PlatformParameterSyncUpWorker::class.java, 12, TimeUnit.HOURS)
```
I am wondering if we should put the number 12 as a dagger constant so that we can change it easily

Arjupta replied:
Yeah it is a good idead but just to make confirm this it will mean introducing a new platform parameter in the codebase.

vinitamurthi commented on domain/src/main/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorker.kt:
```diff
@@ -0,0 +1,92 @@
+    }
+  }
+
+  // It's used to parse the Map of Platform Parameters to a List of Platform Parameter
```
```suggestion
  // Parses a map of platform parameter values into a [List<PlatformParameter>]. If the parameters are not of type String, Int, or Boolean, this function fails with an [IllegalArgumentException].
```

Arjupta replied:
Done

vinitamurthi commented on domain/src/main/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorker.kt:
```diff
@@ -0,0 +1,92 @@
+    return platformParameterList
+  }
+
+  /** Extracts platform parameters from the remote service and store them in the cache store */
```
Change: store -> stores

Arjupta replied:
Done

vinitamurthi commented on domain/src/test/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorkerTest.kt:
```diff
@@ -0,0 +1,343 @@
+  }
+
+  @Test
+  fun testSyncUpWorker_previousDatabaseIsEmpty_refreshPlatformParameters_verifyCachedValues() {
```
In this case, the new values will be set in the cache right? If so change the test name to `testSyncUpWorker_previousDatabaseIsEmpty_refreshPlatformParameters_verifyCachedValuesAreSet`

Arjupta replied:
Going out of 100 max characters. Also see the new names if they look fine to you.

vinitamurthi commented on domain/src/test/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorkerTest.kt:
```diff
@@ -0,0 +1,343 @@
+  }
+
+  @Test
+  fun testSyncUpWorker_previousDatabaseIsNotEmpty_refreshPlatformParameters_verifyCachedValues() {
```
Change to `testSyncUpWorker_previousDatabaseIsNotEmpty_refreshPlatformParameters_verifyCachedValuesAreUpdated`

Arjupta replied:
Going out of 100 max characters. Also see the new names if they look fine to you.

vinitamurthi commented on domain/src/test/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorkerTest.kt:
```diff
@@ -0,0 +1,343 @@
+      expectedTestIntegerParameter
+    )
+  }
+
```
Add test cases for when:
1) There is a value for a flag in the database already, but on refresh a new value for the flag is available that needs to be updated in the cache (i.e. overriding new values)
2) There is a failure in refreshing because of the network response sending a value that is not String,Int, or Boolean so the work manager task fails
3) We request the wrong version name while trying to fetch new data

Arjupta replied:
Updated the test cases kindly take a look.

Arjupta committed:
Added test files in the kdoc-exemption

Arjupta committed:
Corrected the import statement for WorkManagerConfigurationModule

Arjupta committed:
Updating with Base PR

Arjupta committed:
KDoc exemption related changes

Arjupta committed:
Changed the worker test to include exception checking

Arjupta committed:
Nit changes related to KDoc and BUILD.bazel

Arjupta committed:
Nit change of Reference Url in PlatformParameterService

Arjupta committed:
Merge branch 'develop' into platform-parameter-hooking-backend

Arjupta committed:
Merge branch 'platform-parameter-hooking-backend' into platform-parameter-work-manager

Arjupta committed:
Nit changes related to KDoc, BUILD.bazel and WorkMangerConfigurationModule import

Arjupta committed:
Added NetworkModule in MarkTopicsCompletedFragmentTest

Arjupta committed:
Failing checks related to WorkManager imports and Bazel dependency over testing/network

Arjupta commented:
Hey @fsharpasharp can you help me with the Bazel lint issue. Not able to figure out why it is failing.

Arjupta committed:
Trying to fix Bazel failure

Arjupta committed:
Merge branch 'develop' into platform-parameter-work-manager

Arjupta committed:
Nit changes in KDoc exemption file, BUILD.bazel and WorkManagerImport

Arjupta committed:
Suggested Changes for new Param,TestCases and KDoc

- New SyncUpTimePeriod Parameter
- New Test Case for Empty Response in WorkerTests
- Shifted SplashScreenWlecomeMsgServerValue to IntegrationTest
- Few KDoc changes

Arjupta committed:
Nit KDoc change + NetworkModule in TestComponent

Arjupta committed:
Bazel reformating for lint check

Arjupta committed:
Updating with develop

BenHenning commented:
Hi. Will need to look at this tomorrow.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorker.kt:
```diff
@@ -0,0 +1,112 @@
+    private val exceptionsController: ExceptionsController,
+    @BackgroundDispatcher private val backgroundDispatcher: CoroutineDispatcher
+  ) {
+    /** A function that returns an instance of [PlatformParameterSyncUpWorker]. */
```
```suggestion
    /** Returns new instances of [PlatformParameterSyncUpWorker]. */
```

Arjupta replied:
Done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorker.kt:
```diff
@@ -0,0 +1,112 @@
+  // are not of type String, Int or Boolean this function fails with an [IllegalArgumentException].
+  private fun parseNetworkResponse(response: Map<String, Any>): List<PlatformParameter> {
+    val platformParameterList: MutableList<PlatformParameter> = mutableListOf()
+    for (entry in response.entries) {
```
Could we call ``map{}`` on the entries collection instead of using a mutable map?

Arjupta replied:
Yeah. Done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorker.kt:
```diff
@@ -0,0 +1,112 @@
+  }
+
+  /** Extracts platform parameters from the remote service and stores them in the cache store */
+  private fun refreshPlatformParameters(): Result {
```
Could this be marked as a suspend function to make it a bit more clear that it's running within a coroutine?

Arjupta replied:
Two problems with this thing.
1. refresh function is not doing any task in async format so suspend is not needed
2. the execute method over platformParameterService gives a warning message as can be seen in this image

![Untitled](https://user-images.githubusercontent.com/64064110/128905108-8a39aeda-7516-4bd6-b09f-efb3fc58f9ce.png)

BenHenning replied:
TIL but apparently we can change the Retrofit interface to generate a suspend function & return the direct value: https://stackoverflow.com/a/64139147. That's probably the best choice here (though you *may* run into testing issues--please let me know if that's the case).

Arjupta replied:
With your decription, this idea looks good but consider these points - 
- the `refreshPlatformParameter` is a normal function, this means that the control will not return to the `doWork` function untill and unless the execution of `refreshPlatformParameter` is completed.
- as we call the execute method over the `platformParameterService`, it will make sure that the network request is performed synchronously, this means that our `refreshPlatformParameter` function will be blocked to execute further untill and unless we get a response/failure.

So we actually don't need to use suspend for `refreshPlatformParameter` because the current code is working synchrounously only. If we will follow the suggestion to make the `getPlatformParameterByVersion` to be suspend and return a Response\<Map\<String,Any\>\>, then we will need to find a new way to test these suspend function calls in the `PlatformParameterServiceTest`, also we will need to have changes at the places were `MockPlatformParameterService` is getting used. 

I would say we can keep this thing as a completely new Issue to convert the network tests to use Response instead of Calls. (ie. Future Task) as the milestone deadline is coming closer and I have one more PR stacked over this PR, which needs to be merged.

BenHenning replied:
Thanks for looking into this. We do have #2951 tracking related items, so cc'ing it here since this is a useful context to capture.

That being said, I'm not sure that I fully understand your concerns. The fact that ``refreshPlatformParameter`` is a regular function seems a bit moot since it's called in a suspend function (doWork), so it seems like we could easily make it a suspend function. Regarding the blocking behavior, that's actually the problem I think we need to try and avoid. Coroutines are never supposed to be blocked.

That being said, this seems workable for the needs that we currently have & given we don't have better infrastructural support in the app to interop with Retrofit (yet).

BenHenning commented on domain/src/main/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorker.kt:
```diff
@@ -0,0 +1,112 @@
+      if (platformParameterList.isEmpty()) {
+        throw IllegalArgumentException(EMPTY_RESPONSE_EXCEPTION_MSG)
+      }
+      platformParameterController.updatePlatformParameterDatabase(platformParameterList)
```
Should we be observing the DataProvider and/or Deferred to make sure the update was successful? Otherwise we might lose syncs if the fetch succeeds, but the save fails.

Arjupta replied:
Understood. Have implemented the function in a different way now. PTAL

BenHenning replied:
Was this completed? In the latest version I don't see the class actually using the returned DataProvider.

Arjupta replied:
I thought that tests are passing hence the functions are working fine. Added an observerForever over here, not sure about it PTAL.

Arjupta replied:
Just to let you know, after adding observeForever on the dataProvider (which was converted to LiveData after `toLiveData( )`) some of the tests in the PlatformParameterWorkerTest started failing because of this error -> Cannot invoke observeForever on a background thread.

BenHenning replied:
We shouldn't ever use LiveData in the domain layer since it's meant to be a UI-only construct (the lack of Lifecycle being a key reason).

Could we return a Deferred or DataProvider and observe/fetch from that, instead? It probably means this does need to be a suspend function, but that's presumably fine given the execution context.

BenHenning replied:
@ARJUPTA this appears to be unresolved still.

Arjupta replied:
@BenHenning I don't know how but I have changed the code to this format.
![Screenshot 2021-08-19 114619](https://user-images.githubusercontent.com/64064110/130017669-8e1579ae-fb1d-4e9e-9ddc-fab08338f66a.png)
 The code that you are reffering to is outdated one

BenHenning replied:
Yeah, not sure what happened here. The latest implementation looks solid.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorker.kt:
```diff
@@ -0,0 +1,112 @@
+    const val WORKER_TYPE_KEY = "worker_type_key"
+  }
+
+  override suspend fun doWork(): Result {
+    return when (inputData.getString(WORKER_TYPE_KEY)) {
+      PLATFORM_PARAMETER_WORKER -> withContext(backgroundDispatcher) { refreshPlatformParameters() }
+      else -> Result.failure()
+    }
+  }
```
How often will this run per its current configuration?

Arjupta replied:
Right now we have kept it to a default value of once in every 12 Hours. We can change this time period too, if we make it a runtime parameter (add it to backend - future task).

BenHenning commented on utility/src/main/java/org/oppia/android/util/platformparameter/PlatformParameterConstants.kt:
```diff
@@ -15,4 +15,9 @@ annotation class SplashScreenWelcomeMsg
-val SPLASH_SCREEN_WELCOME_MSG_VALUE = true
+
+@Qualifier
+annotation class SyncUpWorkerTimePeriod
+
+val SYNC_UP_WORKER_TIME_PERIOD = "sync_up_worker_time_period"
+val SYNC_UP_WORKER_TIME_PERIOD_DEFAULT_VALUE = 12
```
Please indicate in the name the expected unit (e.g. seconds, minutes, etc.) since that's really important to ensure don't catastrophically reconfigure this in the future.

Arjupta replied:
Done

BenHenning reviewed (changes requested):
Thanks @ARJUPTA. Reviewed all my codeowner files except the tests (so far). Left some clarification questions--PTAL.

jcqli commented on domain/src/test/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorkerTest.kt:
```diff
@@ -0,0 +1,461 @@
+
+  @Test
+  fun testSyncUpWorker_databaseIsEmpty_getCorrectPlatformParameters_verifyValuesAreCached() {
+    // setup versionName to get correct network response from mock platform parameter service
```
nit: `// Set up versionName to get correct network response from mock platform parameter service.` Ditto elsewhere -- writing these as full sentences makes it clearer that this comment is meant to be here. Nice job adding these comments, though, as it makes it very clear what you're testing for!

Arjupta replied:
Done. Thanks

jcqli reviewed (approved):
A couple nits, but otherwise LGTM!

rt4914 reviewed (approved):
Approving for code-owner files. Thanks.

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/administratorcontrols/AdministratorControlsActivityTest.kt:
```diff
@@ -704,7 +705,7 @@ class AdministratorControlsActivityTest {
       FirebaseLogUploaderModule::class, FakeOppiaClockModule::class, PracticeTabModule::class,
       DeveloperOptionsStarterModule::class, DeveloperOptionsModule::class,
-      ExplorationStorageModule::class
+      ExplorationStorageModule::class, NetworkModule::class
```
Could you help me understand why are we adding this?

anandwana001 replied:
@jcqli Do you think this is right? 
I am not seeing it right by importing data module stuff in the app module. 

@rt4914 Could you suggest a better architecture around here? 

@ARJUPTA Could you please add the error log here when we run the test without it?

jcqli replied:
Hm thanks for catching this I'm also curious. Does the `WorkManagerConfigurationModule` need the `NetworkModule`? Was there an error without it?

Arjupta replied:
@jcqli @anandwana001 have replied to your doubt here -> https://github.com/oppia/oppia-android/pull/3606#discussion_r687384854

Arjupta committed:
Suggested change for KDoc, Refactor & DataProvider

- Changed the Time Period Parameter Name
- Made updatePlaformParameter to return a dataprovider
- KDoc changes

Arjupta commented on domain/src/main/java/org/oppia/android/domain/workmanager/WorkManagerConfigurationModule.kt:
```diff
@@ -0,0 +1,26 @@
+@Module
+class WorkManagerConfigurationModule {
+
+  @Singleton
+  @Provides
+  fun provideWorkManagerConfiguration(
+    logUploadWorkerFactory: LogUploadWorkerFactory,
+    platformParameterSyncUpWorkerFactory: PlatformParameterSyncUpWorkerFactory
+  ): Configuration {
```
> https://github.com/oppia/oppia-android/pull/3606#discussion_r686406727

@jcqli @anandwana001 here we have the new `WorkManagerConfigurationModule `which will be injected with `PlatformParameterSyncUpWorkerFactory`. This Factory class requires a will inject the `PlatformParameterSyncUpWorker ` which will directly inject `PlatformParameterService` instance hence we require `NetworkModule ` to be present everywhere the `WorkManagerConfigurationModule `is getting used

jcqli replied:
Ack, I see now. Hm this seems like the correct implementation so I'll defer to @anandwana001 and @rt4914 if they have more perspective on where this could be change. That said, it's being added to an app test so it may not end up being in importing into the app module in prod.

Arjupta replied:
Sure that is true

Arjupta committed:
Updating with develop

Arjupta committed:
Correcting the work manager imports for new files

anandwana001 reviewed (approved):
The need for NetworkModule sounds ok to me. But need to look into more detail to check what other architecture approaches we can use to keep our app module test independent from data module dependencies. 

Approving PR for now as this is something that can be done outside the project, I guess. I will take a look more on this today to solve how it can be a better architect but solve requires @rt4914 and @Sarthak2601 to look here as they worked earlier with similar cases.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/platformparameter/PlatformParameterController.kt:
```diff
@@ -42,25 +46,40 @@ class PlatformParameterController @Inject constructor(
   }
 
+  /**
+   * Execute and Transforms the [Deferred] task into an [AsyncResult].
```
```suggestion
   * Execute and Transforms the [Deferred] task into an [AsyncResult].
```
```suggestion
   * Executes and transforms the [Deferred] task into an [AsyncResult].
```

Arjupta replied:
Done

Arjupta replied:
Done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/platformparameter/PlatformParameterController.kt:
```diff
@@ -42,25 +46,40 @@ class PlatformParameterController @Inject constructor(
+   */
+  private suspend fun getDeferredResult(deferred: Deferred<Boolean>): AsyncResult<Any?> {
+    val throwable = Throwable("Failed when storing platform parameter values list")
+    return if (deferred.await())
+      AsyncResult.success(true)
+    else
+      AsyncResult.failed(throwable)
```
Please add braces per https://developer.android.com/kotlin/style-guide#braces.

Arjupta replied:
Done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/platformparameter/PlatformParameterController.kt:
```diff
@@ -42,25 +46,40 @@ class PlatformParameterController @Inject constructor(
+   * @return async result for success or failure after the execution of deferred task
+   */
+  private suspend fun getDeferredResult(deferred: Deferred<Boolean>): AsyncResult<Any?> {
+    val throwable = Throwable("Failed when storing platform parameter values list")
```
Prefer to use IllegalStateException. Throwable should never be used directly.

Arjupta replied:
Understood. Done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/platformparameter/PlatformParameterController.kt:
```diff
@@ -42,25 +46,40 @@ class PlatformParameterController @Inject constructor(
+   * @param platformParameterList list of [PlatformParameter] objects which needs to be cached
+   * @return a [DataProvider] that indicates the success/failure of this update operation
    */
   fun updatePlatformParameterDatabase(
```
Please add tests to verify that the DataProvider returns the correct values at the expected times.

Arjupta replied:
Done

BenHenning commented on utility/src/main/java/org/oppia/android/util/platformparameter/PlatformParameterConstants.kt:
```diff
@@ -19,5 +19,5 @@ val SPLASH_SCREEN_WELCOME_MSG_DEFAULT_VALUE = false
-val SYNC_UP_WORKER_TIME_PERIOD = "sync_up_worker_time_period"
-val SYNC_UP_WORKER_TIME_PERIOD_DEFAULT_VALUE = 12
+val SYNC_UP_WORKER_TIME_PERIOD_IN_HOURS = "sync_up_worker_time_period"
+val SYNC_UP_WORKER_TIME_PERIOD_IN_HOURS_DEFAULT_VALUE = 12
```
Should have KDocs--please add.

Arjupta replied:
Done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorker.kt:
```diff
@@ -55,9 +55,9 @@ class PlatformParameterSyncUpWorker private constructor(
-    for (entry in response.entries) {
-      val platformParameter = PlatformParameter.newBuilder().setName(entry.key)
-      when (val value = entry.value) {
+    response.map {
```
You should use the result of the map (it returns a List) to avoid needing to construct one piece-by-piece. In this way we can avoid using a mutable list.

Arjupta replied:
Oh got it now. Done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorkManagerInitializer.kt:
```diff
@@ -0,0 +1,76 @@
+    )
+    .build()
+
+  private val workRequestForSyncingPlatformParameters = PeriodicWorkRequest
+    .Builder(
+      PlatformParameterSyncUpWorker::class.java,
+      workRequestRepeatInterval.value.toLong(),
+      TimeUnit.HOURS
+    )
+    .addTag(PlatformParameterSyncUpWorker.TAG)
+    .setInputData(workerTypeForSyncingPlatformParameters)
+    .setConstraints(platformParameterSyncUpWorkerConstraints)
+    .build()
```
```suggestion
  private val workRequestForSyncingPlatformParameters =
    PeriodicWorkRequest.Builder(
      PlatformParameterSyncUpWorker::class.java,
      workRequestRepeatInterval.value.toLong(),
      TimeUnit.HOURS
    )
    .addTag(PlatformParameterSyncUpWorker.TAG)
    .setInputData(workerTypeForSyncingPlatformParameters)
    .setConstraints(platformParameterSyncUpWorkerConstraints)
    .build()
```

Arjupta replied:
Done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorkManagerInitializer.kt:
```diff
@@ -0,0 +1,76 @@
+
+  /** Returns the time interval of periodic work request enqueued to sync-up platform parameters. */
+  @SuppressLint("RestrictedApi")
+  fun getSyncUpWorkerTimePeriod(): Long = workRequestForSyncingPlatformParameters
+    .workSpec.intervalDuration
```
```suggestion
  fun getSyncUpWorkerTimePeriod(): Long =
    workRequestForSyncingPlatformParameters.workSpec.intervalDuration
```

Wrapping always starts at the higher syntactical element (in this case the ``=`` before the ``.``).

Arjupta replied:
Done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorkManagerInitializer.kt:
```diff
@@ -0,0 +1,76 @@
+ * Enqueues unique periodic work requests for fetching and caching latest platform parameter values
+ * from the remote service on application creation.
+ */
+@Singleton
```
Is this injected in more than one place? Since it doesn't contain state, we don't actually need to scope it.

Arjupta replied:
No its not injected more than once. Done

BenHenning commented on domain/src/test/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorkManagerInitializerTest.kt:
```diff
@@ -0,0 +1,239 @@
+    val workInfoList = workManager.getWorkInfosByTag(PlatformParameterSyncUpWorker.TAG).get()
+    // There should be only one such work request having "PlatformParameterSyncUpWorker.TAG" tag
+    assertThat(workInfoList.size).isEqualTo(1)
+    // Match the Id of this work request with the Id of another work request which was enqueued by
```
```suggestion
    // Match the ID of this work request with the ID of another work request which was enqueued by
```

Arjupta replied:
Done

BenHenning commented on domain/src/test/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorkManagerInitializerTest.kt:
```diff
@@ -0,0 +1,239 @@
+    assertThat(workInfoList.size).isEqualTo(1)
+    // Match the Id of this work request with the Id of another work request which was enqueued by
+    // PlatformParameterSyncUpWorkManagerInitializer
+    assertThat(workInfoList[0].id).isEqualTo(enqueuedSyncUpWorkRequestId)
```
I think you need to reverse the values here. For Truth, the syntax is: assertThat(subject).isEqualTo(expectedValue)

Arjupta replied:
Done

BenHenning commented on domain/src/test/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorkManagerInitializerTest.kt:
```diff
@@ -0,0 +1,239 @@
+    testCoroutineDispatchers.runCurrent()
+
+    val syncUpWorkerTimePeriodInMs = syncUpWorkManagerInitializer.getSyncUpWorkerTimePeriod()
+    val syncUpWorkerTimePeriodInHours = syncUpWorkerTimePeriodInMs / (60 * 60 * 1000)
```
```suggestion
    val syncUpWorkerTimePeriodInHours = TimeUnit.MILLIS.toHours(syncUpWorkerTimePeriodInMs)
```

Arjupta replied:
Done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorkManagerInitializer.kt:
```diff
@@ -0,0 +1,76 @@
+    )
+  }
+
+  /** Returns the Worker [Constraints] set for the platform parameter sync-up work requests. */
+  fun getSyncUpWorkerConstraints(): Constraints = platformParameterSyncUpWorkerConstraints
+
+  /** Returns the [UUID] of the work request that is enqueued to sync-up platform parameters. */
+  fun getSyncUpWorkRequestId(): UUID = workRequestForSyncingPlatformParameters.id
+
+  /** Returns the [Data] that goes into the work request enqueued to sync-up platform parameters. */
+  fun getSyncUpWorkRequestData(): Data = workerTypeForSyncingPlatformParameters
```
Are these used outside tests? If not, we should find another way to verify this state rather than creating methods specifically for the purpose of testing internal class state.

Arjupta replied:
As the `LogUploadWorkManagerInitializer` (which was created by Sarthak earlier) was having all these methods defined in the prod class onyl  therefore I made these functions here. Now shifting these functions to defined in the test. PTAL

BenHenning replied:
Per the latest changes, that's not quite what I meant. The issue is that we shouldn't expose implementation details of classes only for testing purposes (this essentially "punches holes" in APIs and can result in unintended side effects like people calling into these internal states outside tests when they shouldn't be).

While the functions have been moved to tests, it required you exposing internal state (the vals) which is potentially problematic.

It seems like we can potentially only expose workRequestForSyncingPlatformParameters since that contains references to the other two. Is that possible?

Further, is it possible to retrieve workRequestForSyncingPlatformParameters from the work manager rather than exposing this? If not, we should at least mark the val as VisibileForTesting (with private default visibility).

Arjupta replied:
I have tried out on my end to extract these properties from `workRequestForSyncingPlatformParameters` and `WorkInfo` object in the test, but was not successfull. Therefore I am shifting the functions back to the prod file and marking it visible for testing. PTAL

BenHenning replied:
The latest changes seem reasonable--thanks!

BenHenning commented on domain/src/test/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorkerTest.kt:
```diff
@@ -0,0 +1,461 @@
+    // Controller in the form of a Map, therefore verify the retrieved values from that Map.
+    val platformParameterMap = platformParameterSingleton.getPlatformParameterMap()
+    assertThat(platformParameterMap).isNotEmpty()
+    assertThat(platformParameterMap).containsEntry(
```
This implies "isNotEmpty" so the check above can probably be removed.

Arjupta replied:
Thought that it might be better to do so. Updated the code, done

BenHenning commented on domain/src/test/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorkerTest.kt:
```diff
@@ -0,0 +1,461 @@
+    // Controller in the form of a Map, therefore verify the retrieved values from that Map.
+    val platformParameterMap = platformParameterSingleton.getPlatformParameterMap()
+    assertThat(platformParameterMap).isNotEmpty()
+    assertThat(platformParameterMap).containsEntry(
+      TEST_STRING_PARAM_NAME,
+      expectedTestStringParameter
+    )
```
If this can fit on one line, prefer that it does. Ditto elsewhere.

Arjupta replied:
It doesn't fit in one line, though I have tried something subtle, PTAL

BenHenning commented on domain/src/test/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorkerTest.kt:
```diff
@@ -0,0 +1,461 @@
+  }
+
+  @Test
+  fun testSyncUpWorker_databaseIsEmpty_getWrongPlatformParameters_verifyWorkerCrashes() {
```
Will this scenario result in the app crashing?

Arjupta replied:
No only the Worker will crash and the exception will be logged with the help of ExceptionLogger. App will work normally

BenHenning commented on domain/src/test/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorkerTest.kt:
```diff
@@ -0,0 +1,461 @@
+    // Values retrieved from Cache store will be sent to Platform Parameter Singleton by the
+    // Controller in the form of a Map, therefore verify the retrieved values from that Map.
+    val platformParameterMap = platformParameterSingleton.getPlatformParameterMap()
+    assertThat(platformParameterMap).isNotEmpty()
```
Does this actually verify that the values aren't updated? It seems to me that this could accidentally pass in a lot of other scenarios. Can we make this a bit more specific to ensuring that the values actually didn't change?

Arjupta replied:
Yeah sure, Done. If possible can you share some cases at which this would accidentally pass.

BenHenning reviewed (changes requested):
Thanks @ARJUPTA. Took another pass (this time full since I also covered tests per my codeowner files).

Arjupta committed:
Suggested Nit changes by Ben

Arjupta committed:
More Nit changes

Arjupta committed:
Adding the tests for DataProvider we get from update operation for platform parameter database

Arjupta committed:
Nit KDoc changes

Arjupta committed:
Updating with develop

vinitamurthi commented on domain/src/main/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorker.kt:
```diff
@@ -0,0 +1,122 @@
+    }
+  }
+
+  // Parses a map of platform parameter values into a [List<PlatformParameter>]. If the parameters
+  // are not of type String, Int or Boolean this function fails with an [IllegalArgumentException].
```
Make these into KDoc comments (i.e. `/** ..... */`)

Arjupta replied:
Done

vinitamurthi commented on domain/src/main/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorker.kt:
```diff
@@ -0,0 +1,122 @@
+    }
+  }
+
+  // Synchronously executes the network request to get platform parameters from the Oppia backend
```
Make these into KDoc comments (i.e. `/** ..... */`)

Arjupta replied:
Done

vinitamurthi commented on domain/src/test/java/org/oppia/android/domain/platformparameter/PlatformParameterControllerTest.kt:
```diff
@@ -102,6 +108,60 @@ class PlatformParameterControllerTest {
+  }*/
+
+/*  @Test
+  fun testController_existingDatabase_updateDatabaseWithNullList_cachingOperationFails() {
+    // Simulate that previous app already has cached platform parameter values in cache store.
+    executeInPrevious { testComponent ->
+      testComponent.getPlatformParameterController().updatePlatformParameterDatabase(
+        mockPlatformParameterList
+      ).toLiveData().observeForever(mockObserverForAny)
+      testComponent.getTestCoroutineDispatchers().advanceUntilIdle()
+    }
+
+    // Create the application after previous arrangement to simulate a re-creation.
+    setUpTestApplicationComponent()
+    reset(mockObserverForAny)
+    // Update the platform parameter with a null list that should fail as per the implementation.
+    platformParameterController.updatePlatformParameterDatabase(null)
+      .toLiveData().observeForever(mockObserverForAny)
+    testCoroutineDispatchers.advanceUntilIdle()
+
+    verify(mockObserverForAny, atLeastOnce()).onChanged(captorForAny.capture())
+    val asyncResult = captorForAny.value
+
+    // Check that the operation failed and match the exception message.
+    assertThat(asyncResult.isFailure()).isTrue()
+    assertThat(asyncResult.getErrorOrNull()?.message)
+      .isEqualTo(PlatformParameterController.EXCEPTION_MSG_FOR_CACHING_FAILURE)
+
+    // Fetch the values from database to verify they are not updated.
+    platformParameterController.getParameterDatabase().toLiveData().observeForever(mockUnitObserver)
+    testCoroutineDispatchers.advanceUntilIdle()
+
+    // Verify the fetch operation was successful.
+    verify(mockUnitObserver, atLeastOnce()).onChanged(unitCaptor.capture())
+    assertThat(unitCaptor.value.isSuccess()).isTrue()
+
+    // Verify the map injected into platformParameterSingleton after successful fetch operation.
+    assertThat(platformParameterSingleton.getPlatformParameterMap()).isNotEmpty()
+    verifyEntriesInsidePlatformParameterMap(platformParameterSingleton.getPlatformParameterMap())
+  }*/
```
Is this supposed to be commented out?

Arjupta replied:
These tests were supposed to be removed. I commented and minimized them, therefore missed to delete this comment before commiting. Removed now.

vinitamurthi reviewed (changes requested):
Overall LGTM, just a few things. Kept my review as blocking mainly because of the commented out tests in one of the files. LEt me know once you change that and I will take a look again!

Arjupta committed:
KDoc changes and deleting commented out code

Arjupta committed:
Updating with develop

BenHenning commented on domain/src/main/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorkManagerInitializer.kt:
```diff
@@ -11,45 +10,45 @@ import androidx.work.WorkManager
 
-  private val platformParameterSyncUpWorkerConstraints = Constraints.Builder()
+  /** [Constraints] for platform parameter sync up work request. */
+  val platformParameterSyncUpWorkerConstraints = Constraints.Builder()
```
Here & below: are these immutable structures? If not, we need to be really careful about exposing them like this.

Arjupta replied:
No they are not. Changed the approach now. PTAL

BenHenning commented on utility/src/main/java/org/oppia/android/util/platformparameter/PlatformParameterConstants.kt:
```diff
@@ -16,8 +16,21 @@ annotation class SplashScreenWelcomeMsg
+ * [PlatformParameterSyncUpWorker] will run again.
+ */
 @Qualifier
 annotation class SyncUpWorkerTimePeriod
```
If this is in hours, please include that in the name (e.g. ``SyncUpWorkerTimePeriodHours`` to make it clear at the injection site).

Arjupta replied:
Done

ARJUPTA replied:
Done

BenHenning commented on domain/src/test/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorkManagerInitializerTest.kt:
```diff
@@ -160,6 +163,28 @@ class PlatformParameterSyncUpWorkManagerInitializerTest {
+  }
+
+  /** Returns the time interval of periodic work request enqueued to sync-up platform parameters. */
+  @SuppressLint("RestrictedApi")
```
Why is this needed? Please include a brief line comment next to the suppression to justify it, e.g.:

```kotlin
@SupressLint("RestrictedApi") // Needed because X, but we can suppress because Y.
```

Arjupta replied:
Done

BenHenning commented on domain/src/test/java/org/oppia/android/domain/platformparameter/PlatformParameterControllerTest.kt:
```diff
@@ -162,6 +168,39 @@ class PlatformParameterControllerTest {
   }
 
+  @Test
+  fun testController_noPreviousDatabase_performUpdateOperation_checkAsyncResult() {
```
```suggestion
  fun testController_noPreviousDatabase_performUpdateOperation_returnsSuccess() {
```

Ditto below (better to be as specific about the expected outcome as we can, particularly how it relates back to the operation being run).

Arjupta replied:
Done

ARJUPTA replied:
Done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorker.kt:
```diff
@@ -51,35 +53,45 @@ class PlatformParameterSyncUpWorker private constructor(
-  // Parses a map of platform parameter values into a [List<PlatformParameter>]. If the parameters
-  // are not of type String, Int or Boolean this function fails with an [IllegalArgumentException].
+  /**
+   * Parses a map of platform parameter values into a [List<PlatformParameter>]. If the parameters
+   * are not of type String, Int or Boolean this function fails with an [IllegalArgumentException].
```
```suggestion
   * Parses a map of platform parameter values into a [List<PlatformParameter>]. Parameters must be
   * of type String, Int or Boolean.
```

We actually don't want to indicate the type of exception in this case since it's a runtime exception & not meant to be caught.

Arjupta replied:
Understood. Done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/platformparameter/PlatformParameterController.kt:
```diff
@@ -29,6 +24,19 @@ class PlatformParameterController @Inject constructor(
     RemotePlatformParameterDatabase.getDefaultInstance()
   )
 
+  /** It corresponds to the state of the platform parameter caching process. */
```
```suggestion
  /** The state of the platform parameter caching process. */
```

Arjupta replied:
Done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/platformparameter/PlatformParameterController.kt:
```diff
@@ -29,6 +24,19 @@ class PlatformParameterController @Inject constructor(
 
+  /** It corresponds to the state of the platform parameter caching process. */
+  private enum class PlatformParameterCachingStatus {
+    /** If the deferred contains a [SUCCESS], it corresponds to a successful [AsyncResult]. */
```
Suggest instead: ``Whether the caching operation corresponding to this status was successful.``

Arjupta replied:
Done

BenHenning commented on domain/src/test/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorkManagerInitializerTest.kt:
```diff
@@ -119,7 +122,7 @@ class PlatformParameterSyncUpWorkManagerInitializerTest {
-    assertThat(syncUpWorkManagerInitializer.getSyncUpWorkRequestData()).isEqualTo(
+    assertThat(getSyncUpWorkRequestData()).isEqualTo(
       workerTypeForSyncingUpParameters
     )
```
Can this fit on one line? If so, prefer to update (in general, try to keep everyone on one line if it fits vs. adding extra wrapping).

Arjupta replied:
Done. Sure

BenHenning commented on domain/src/test/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorkerTest.kt:
```diff
@@ -327,17 +318,24 @@ class PlatformParameterSyncUpWorkerTest {
-    assertThat(platformParameterMap).isNotEmpty()
+
+    // Previous String Platform Parameter is still same in the Database.
+    assertThat(platformParameterMap)
+      .containsEntry(TEST_STRING_PARAM_NAME, expectedTestStringParameter)
+
+    // Previous Integer Platform Parameter is still same in the Database.
+    assertThat(platformParameterMap)
+      .containsEntry(TEST_INTEGER_PARAM_NAME, defaultTestIntegerParameter)
```
Suggest using containsExactly() to fully verify the map & ensure nothing else is included.

Arjupta replied:
As the map is supposed to have more than two platforms here so containsExactly will not serve the purpose.

BenHenning reviewed (commented):
Thanks @ARJUPTA. Followed up on all recent changes/comments. PTAL.

BenHenning commented:
Also @ARJUPTA I suggest syncing with the latest develop changes & updating this PR accordingly.

Arjupta committed:
Resolving conflicts

Arjupta committed:
Suggested changes for KDoc and VisibleForTesting functions

Arjupta committed:
Added NetworkModule in ForceNetworkTypeTests

Arjupta committed:
Add NetworkConnectionDebugUtilModule in domain/BUILD.bazel

BenHenning commented on domain/src/main/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorkManagerInitializer.kt:
```diff
@@ -58,4 +61,27 @@ class PlatformParameterSyncUpWorkManagerInitializer @Inject constructor(
+    return workerTypeForSyncingPlatformParameters
+  }
+  /** Returns the time interval of periodic work request enqueued to sync-up platform parameters. */
+  @SuppressLint("RestrictedApi") /** Needed because getWorkSpec is restricted to library group but we can suppress for testing purpose. */
```
This should not be a KDoc comment, and it's longer than the 100 line limit. Suggest:

```suggestion
  @SuppressLint("RestrictedApi") // getWorkSpec is restricted; suppression is fine for tests.
```

Arjupta replied:
Done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorkManagerInitializer.kt:
```diff
@@ -58,4 +61,27 @@ class PlatformParameterSyncUpWorkManagerInitializer @Inject constructor(
+  fun getSyncUpWorkRequestData(): Data {
+    return workerTypeForSyncingPlatformParameters
+  }
+  /** Returns the time interval of periodic work request enqueued to sync-up platform parameters. */
```
Newline before this.

Arjupta replied:
Done

BenHenning reviewed (changes requested):
Thanks @ARJUPTA. Just a couple of new comments, and one existing thread that I think hasn't yet been resolved. Other than these last few comments the PR LGTM.

PTAL.

Arjupta committed:
Suggested Nit changes

Arjupta committed:
Resolving conflicts

Arjupta committed:
Merge branch 'develop' into platform-parameter-work-manager

Arjupta committed:
Bazel check failure due to NetworkConfigProdModule

Arjupta committed:
Adding NetworkConfigProdModule into domain tests

Arjupta committed:
Adding NetworkConfigProdModule in StateFragmentTest

Arjupta committed:
Added WorkManagerConfigurationModule in OppiaTestApplication

Arjupta committed:
Injected baseUrl into SyncuUpWorkerTests

Arjupta committed:
Correcting the imports in TestApplicationComponent

BenHenning reviewed (approved):
Thanks @ARJUPTA. This LGTM!

BenHenning commented:
Given other threads seem resolved & everyone has approved, merging this.
